### DISCUSSION TITLE
Running examples on CPU

### DISCUSSION THREAD

**Discussion by uturuncoglu** [NVIDIA/physicsnemo#427] [Q&A]
I am trying to get familiar to `NVIDIA/modulus` but I have no access to any GPU. Is it possible to run the tests (like `examples/cfd/darcy_fno`) on CPU? I have just download the Docker container and started with the `docker run --shm-size=1g --ulimit memlock=-1 --ulimit stack=67108864 --rm -it nvcr.io/nvidia/modulus/modulus:24.01 bash` command since I have no GPU access on my local laptop. It seems I could run very simple `torch` code which is available in the documentation. But I am getting runtime error like `RuntimeError: "cuda" device requested but CUDA is not supported by the hardware or driver` from `darcy_fno` test. I think I need to modify the code to work on CPU (probably `DistributedManager` in `darcy_trainer`) but don't know how. Anyway, if you guide me about that that would be great.

**Comment by sudoCommit** [comment]
Hi  @uturuncoglu,
I checked the [train_fno_darcy.py](https://github.com/NVIDIA/modulus-launch/blob/main/examples/cfd/darcy_fno/train_fno_darcy.py) and there are few places were we check for device, like in [model](https://github.com/NVIDIA/modulus-launch/blob/main/examples/cfd/darcy_fno/train_fno_darcy.py#L74) and this `dist` is defined in [DistributedManager](https://github.com/NVIDIA/physicsnemo/blob/main/physicsnemo/distributed/manager.py#L104) where we have device agnostic code, so it should use CPU when GPU is not available.

**I guess I found the culprit**, its when [dataloader](https://github.com/NVIDIA/modulus-launch/blob/main/examples/cfd/darcy_fno/train_fno_darcy.py#L85) is defined you need to pass `device` otherwise it'll default to GPU in [Darcy2D](https://github.com/NVIDIA/physicsnemo/blob/main/physicsnemo/datapipes/benchmarks/darcy.py#L106)
So updating the code it should work.
```
dataloader = Darcy2D(
    resolution=cfg.training.resolution,
    batch_size=cfg.training.batch_size,
    normaliser=normaliser,
    device=torch.device("cpu")      # or device=dist.device
)
```


Optional if above dosen't work
1 way to try is when you launch the container, try explicitly blocking GPU visibility
```
docker run \
  --shm-size=1g \
  --ulimit memlock=-1 \
  --ulimit stack=67108864 \
  --rm -it \
  -e CUDA_VISIBLE_DEVICES="" \
  nvcr.io/nvidia/modulus/modulus:24.01 bash
```
